src/poetry/core/packages/package.py

```diff
@@ -559,7 +559,11 @@ def clone(self: T) -> T:
         return clone
 
     def __hash__(self) -> int:
-        return super().__hash__() ^ hash(self._version)
+        return (
+            super().__hash__()
+            ^ hash(self._version)
+            ^ hash(frozenset(self.extras.keys()))
+        )
 
     def __eq__(self, other: object) -> bool:
```

Comment:
_"Hashable objects which compare equal must have the same hash value."_ ([source](https://docs.python.org/3/glossary.html#term-hashable)) That's not fulfilled here. Since `__eq__` relies on `is_same_package_as()` and that can be fulfilled even if `_source_reference` attributes are not equal and `_source_reference` is used in hash.

Reply:
I think, there are two options:

1. Don't use `is_same_package_as()` in `__eq__` and just compare everything that is used in `__hash__`.
2. Only use attributes in `__hash__` that have to be equal according to `is_same_package_as()`. That probably means, we can't use `_source_reference` and `_resolved_source_reference` in hash.

I suppose 2. might be ok. 1. seems to be a breaking change.